keps/sig-api-machinery/3157-watch-list/README.md

```diff
@@ -0,0 +1,1026 @@
+In extreme cases, the server can allocate hundreds of megabytes per request.
+To better visualize the issue let's consider the above graph.
+It shows the memory usage of an API server during a test (see [manual test](#manual-testing-without-the-changes-in-place) section for more details).
+We can see that increasing the number of informers drastically increases the memory consumption of the server.
+Moreover, around 16:40 we lost the server after running 16 informers. During an investigation, we realized that the server allocates a lot of memory for handling LIST requests.
+In short, it needs to bring data from the database, unmarshal it, do some conversions and prepare the final response for the client.
+The bottom line is around O(5*the_response_from_etcd) of temporary memory consumption.
+Neither priority and fairness nor Golang garbage collection is able to protect the system from exhausting memory.
+
+A situation like that is dangerous twofold.
+First, as we saw it could slow down if not fully stop an API server that has received the requests.
+Secondly, a sudden and uncontrolled spike in memory consumption will likely put pressure on the node itself.
+This might lead to thrashing, starving, and finally losing other processes running on the same node, including kubelet.
+Stopping kubelet has serious issues as it leads to workload disruption and a much bigger blast radius.
+Note that in that scenario even clusters in an HA setup are affected.
+
+Worse, in rare cases (see the [Appendix](#appendix) section for more) recovery of large clusters with therefore many kubelets and hence informers for pods, secrets, configmap can lead to a very expensive storm of LISTs.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+- protect kube-apiserver and its node against list-based OOM attacks
+- considerably reduce (temporary) memory footprint of LISTs, down from O(watchers\*page-size\*object-size\*5) to O(watchers\*constant), constant around 2 MB.
+>  Example:
+> 
+>  512 watches of 400mb data: 512*500*2MB*5=2.5TB ↘ 2 GB
+> 
+>  racing with Golang GC to free this temporary memory before being OOM'ed.
+
+- reduce etcd load by serving from watch cache
+- get a replacement for paginated lists from watch-cache, which is not feasible without major investment
+- enforce consistency in the sense of freshness of the returned list
+- be backward compatible with new client -> old server
+- fix the long-standing "stale reads from the cache" issue, https://github.com/kubernetes/kubernetes/issues/59848
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- get rid of list or list pagination
+- rewrite the list storage stack to allow streaming, but rather use the existing streaming infrastructure (watches).
+
+## Proposal
```

Comment:
Here's an alternative:

a) Force everyone to use pagination (e.g. serialize non-paginated lists)
b) The server may decide to reduce the page size based on the size of the response from etcd.
c) The continue token is of course changed to point at the next unseen item.

Reply:
See also: https://github.com/kubernetes/kubernetes/issues/98541

Past discussion: https://github.com/kubernetes/kubernetes/issues/90179